doc/neps/nep-XXXX-SIMD-optimizations.rst

```diff
@@ -0,0 +1,229 @@
+- Infrastructure is provided in the code for abstract intrinsics. The ufunc
+  machinery will be extended using sets of these abstract intrinsics, so that
+  a single ufunc will be expressed as a set of loops, going from a minimal to
+  a maximal set of possibly availabe intrinsics.
+- At compile time, compiler macros and CPU detection are used to turn the
+  abstract intrinsics into concrete intrinsic calls. Any intrinsics not
+  available on the platform, either because the CPU does not support them
+  (and so cannot be tested) or because the abstract intrinsic does not have a
+  parallel concrete intrinsic on the platform will not error, rather the
+  corresponding loop will not be produced and added to the set of
+  possibilities.
+- At runtime, the CPU detection code will further limit the set of loops
+  available, and the optimal one will be chosen for the ufunc.
+
+The current NEP proposes only to use the runtime feature detection and optimal
+loop selection mechanism for ufuncs. Future NEPS may propose other uses for the
+proposed solution.
+
+Usage and Impact
+----------------
+
+The end user will be able to get a list of available intrinsics. Optionally,
+the user may be able to specify which of the loops available at runtime will be
+used, perhaps via an environment variable to enable benchmarking the impact of
+the different loops. There should be no direct impact to naive end users, the
+results of all the loops should be identical to within a small number (1-3)
+ULPs. On the other hand, users with more powerful machines should notice a
+performance boost.
+
+Binary releases - wheels on PyPI and conda packages
+```````````````````````````````````````````````````
+
+The binaries released by this process will be larger since they include all
+possible loops for the architecture. Some packagers may prefer to limit the
+number of loops in order to limit the size of the binaries, we would hope they
+would still support a wide range of families of architectures. Note this
+problem already exists in the Intel MKL offering.
+
+Source builds
+`````````````
+TBD
+- Setting the baseline and set of runtime-dispatchable ISA extensions
+- Behavior when compiler or hardware doesn't support a requested ISA extension
+
+
+How to run benchmarks to assess performance benefits
+````````````````````````````````````````````````````
+
+Adding more code which use intrinsics will make the code harder to maintain.
+Therefore, such code should only be added if it yields a significant
+performance benefit. Assessing this performance benefit can be nontrivial.
+To aid with this, the implementation for this NEP will add a way to select
+which instruction sets can be used at *runtime* via environment variables.
+(name TBD).
+
+
+Diagnostics
+```````````
+
+A new dictionary `__cpu_features__` will be available to python. The keys are
+the available features, the value is a boolean ``True``. Various new private
```

Comment:
This should either:
* Have `False` keys for features that numpy knows about but aren't present
* Be a `set` instead of a `dict`

Reply:
It should contain all CPU features that NumPy supports and the value `True` or `False` depends on the runtime availability.  Mainly used by test files and reports
A runtime dump from gh-13421:
```Python
>>> np.__cpu_features__
{
    'MMX': True, 'SSE2': True, 'SSE3': True, 'SSSE3': True, 'SSE41': True,
    'POPCNT': True, 'SSE42': True, 'AVX': True, 'F16C': True, 'XOP': False,
    'FMA4': False, 'FMA3': True, 'AVX2': True, 'AVX512F': False, 'AVX512CD': False,
    'AVX512ER': False, 'AVX512PF': False, 'AVX5124FMAPS': False, 'AVX5124VNNIW': False,
    'AVX512VPOPCNTDQ': False, 'AVX512VL': False, 'AVX512BW': False, 'AVX512DQ': False,
    'AVX512VNNI': False, 'AVX512IFMA': False, 'AVX512VBMI': False, 'AVX512VBMI2': False,
    'AVX512BITALG': False, 'AVX512_KNL': False, 'AVX512_KNM': False, 'AVX512_SKX': False,
    'AVX512_CLX': False, 'AVX512_CNL': False, 'AVX512_ICL': False,
    'VSX': False, 'VSX2': False, 'VSX3': False,
    'NEON': False, 'NEON_FP16': False, 'NEON_VFPV4': False, 'ASIMD': False,
    'FPHP': False, 'ASIMDHP': False, 'ASIMDDP': False, 'ASIMDFHM': False
}
````

There also another two attributes implemented by gh-13516 :
    -  `__cpu_baseline__` a list of CPU baseline feature names that configured by command argument `--cpu-baseline`
    -  `__cpu_dispatch__` a list of CPU dispatch feature names that configured by command argument `--cpu-dispatch`

these attributes do not contain the exact values that the user provides from command argument
but with the final values after a long process of testing and filtering depending on the user platform and compiler support.

A runtime dump from gh-13516 (the default configuration on x86):
```Python
>>> np.__cpu_baseline__
['SSE', 'SSE2', 'SSE3']

>>> np.__cpu_dispatch__
[
    'SSSE3', 'SSE41', 'POPCNT', 'SSE42', 'AVX', 'F16C', 'XOP', 'FMA4', 
    'FMA3', 'AVX2', 'AVX512F',  'AVX512CD', 'AVX512_KNL', 'AVX512_KNM', 
    'AVX512_SKX', 'AVX512_CLX', 'AVX512_CNL', 'AVX512_ICL'
]
````
The three attrubites are neccerssy in testing process for SIMD features and also for reports.
An example of putting the three attributes in action:

```Bash
NumPy version 1.19.0.dev0+599ac14
NumPy relaxed strides checking option: True
# a simple report printed at the beginning of tests(OpenCV like) from (gh-13516)
# without '*' or '?' that means the feature is part of CPU baseline
# with '*' part of CPU dispatch
# with '?' part of CPU dispatch but the current CPU doesn't support this feature.
NumPy CPU features: SSE SSE2 SSE3 SSSE3* SSE41* POPCNT* SSE42* AVX* F16C* XOP? FMA4? FMA3* AVX2* AVX512F? AVX512CD? AVX512_KNL? AVX512_KNM? AVX512_SKX? AVX512_CLX? AVX512_CNL? AVX512_ICL?
````

Reply:
updating with this information, I misinterpreted the code.